libs/@local/hql/cst/src/value.rs

```diff
@@ -0,0 +1,46 @@
+use alloc::borrow::Cow;
+
+use hql_span::SpanId;
+use json_number::Number;
+
+use crate::{arena, Spanned};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Entry<'arena, 'source> {
+    pub key_span: SpanId,
+
+    pub value: Value<'arena, 'source>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ValueKind<'arena, 'source> {
+    /// Represents a JSON null value
+    Null,
+
+    /// Represents a JSON boolean
+    Bool(bool),
+
+    /// Represents a JSON number, wether integer or floating point
+    Number(Cow<'source, Number>),
+
+    /// Represents a JSON string
+    String(Cow<'source, str>),
+
+    /// Represents a JSON array
+    Array(arena::Vec<'arena, Value<'arena, 'source>>),
+
+    /// Represents a JSON object
+    Object(arena::HashMap<'arena, Cow<'source, str>, Entry<'arena, 'source>>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Value<'arena, 'source> {
+    pub kind: ValueKind<'arena, 'source>,
+    pub span: SpanId,
```

Comment:
I thought we won't have nested spans in values? For arrays and objects that's a quite big nesting, no? Or is this expected to happen quite rarely?

Reply:
I thought we did? Now that we have our own value, it's quite easy to add, and it's really useful in conversion code, and handling values.

You can see that with the duplicate key diagnostic, that wouldn't have been possible otherwise.

Reply:
Yes, I see it's being useful, no worries. My only concern is that resolving takes longer but that is only in the error path so all good 🙂